packages/jupyterlab-manager/src/manager.ts

```diff
@@ -189,29 +198,51 @@ class WidgetManager extends ManagerBase<Widget> implements IDisposable {
       return;
     }
     await this.context.session.ready;
+    if ((this.context.session.kernel as any).handleComms === false) {
+      return;
+    }
     const comm_ids = await this._get_comm_info();
 
-    // For each comm id, create the comm, and request the state.
+    // For each comm id that we do not know about, create the comm, and request the state.
     const widgets_info = await Promise.all(Object.keys(comm_ids).map(async (comm_id) => {
-      const comm = await this._create_comm(this.comm_target_name, comm_id);
-      const update_promise = new Promise<Private.ICommUpdateData>((resolve, reject) => {
-        comm.on_msg((msg) => {
-          put_buffers(msg.content.data.state, msg.content.data.buffer_paths, msg.buffers);
-          // A suspected response was received, check to see if
-          // it's a state update. If so, resolve.
-          if (msg.content.data.method === 'update') {
-            resolve({
-              comm: comm,
-              msg: msg
+      try {
+        await this.get_model(comm_id);
+        // If we successfully get the model, do no more.
+        return;
+      } catch (e) {
+        // If we have the widget model not found error, then we can create the
+        // widget. Otherwise, rethrow the error.
+        if (e.message !== 'widget model not found') {
+          throw e;
+        }
+        const comm = await this._create_comm(this.comm_target_name, comm_id);
+
+        let msgId: string;
+        let p = new PromiseDelegate<Private.ICommUpdateData>();
```

Comment:
`p` is not very descriptive :)
Also, does not seem you are reassigning the variable should you use `const` instead of `let`?

Reply:
Good point. Fix coming right up.